### PR TITLE
Introduce auto detection of format

### LINKED ISSUE #9964: Make logstash Filebeat module use multiple pipelines
Similar to what we've done with the elasticsearch module here: https://github.com/elastic/beats/pull/8852.

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

This PR introduces auto-detection of Logstash's log file format (plaintext or JSON) and calls the appropriate ingest pipeline for parsing.

## Why is it important?

The `logstash` Filebeat module has always has the ability to parse either plaintext or JSON logs emitted by Logstash. Prior to this PR users would need to manually choose a format by specifying the `var.format` configuration setting in their Logstash module configuration. 

With this PR they will no longer need to manually choose the format; the module will auto-detect it for them. This is in line with what we do in the `elasticsearch` Filebeat module.

This change is also a requirement for migration modules to packages (see https://github.com/elastic/package-registry/pull/270#discussion_r393719035).

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] ~I have commented my code, particularly in hard-to-understand areas~
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] ~I have added tests that prove my fix is effective or that my feature works~ Tests already exist.
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues
- Closes elastic/beats#9964

### CODE REVIEW

File: filebeat/module/logstash/log/config/log.yml [hunk 1]
```diff
@@ -5,12 +5,10 @@ paths:
 {{ end }}
 exclude_files: [".gz$"]
 
-{{ if eq .format "plain" }}
 multiline:
-  pattern: ^\[[0-9]{4}-[0-9]{2}-[0-9]{2}
+  pattern: ^(\[[0-9]{4}-[0-9]{2}-[0-9]{2}|{)
```
Comment (mtojek) [inline filebeat/module/logstash/log/config/log.yml:9]: Is it possible to trick this pattern with `{` character? I mean having a plain text file with curly brackets inside.

Comment (ycombinator) [reply filebeat/module/logstash/log/config/log.yml:9]: Yes, it's a good point. If there's a multiline log event where, say, the 2nd line starts with a `{`, then this pattern breaks down. Unfortunately, I'm not really sure how to handle this scenario well.

Comment (mtojek) [reply filebeat/module/logstash/log/config/log.yml:9]: How about extending this pattern to `{"level"` ?

**Comment by ycombinator** [comment]
> The only concerns I have are mixed files (plain text combined together with JSON).

If a plain text log event has JSON anywhere after the first character, it should be handled fine. The problem only comes with plain text log events that have `{` as the first character of a line, which could happen in multiline plaintext events (see our other discussion in the comment about this).

File: filebeat/module/logstash/log/config/log.yml (hunk 1 above)
Comment (ycombinator) [reply filebeat/module/logstash/log/config/log.yml:9]: For JSON-formatted logs, each log line is a JSON object. Being an _object_, I don't want to depend on a specific property, e.g. `level`, being the first one.

**Comment by mtojek** [comment]
I wonder if we can use the fact that a plain text file will always be a plain text file (and also the other way round).

**Comment by mtojek** [comment]
Another idea, possible stupid one.

Try to parse the line as JSON and fallback to plain text.

**Comment by ycombinator** [comment]
> I wonder if we can use the fact that a plain text file will always be a plain text file (and also the other way round).

Yes, this would be ideal, as we don't expect a plaintext log file to suddenly start containing JSON log entries and vice versa. However, that would require some way for libbeat to "tell" the ES ingest pipeline, per-event, that "this event came from a plaintext file" vs. "this event came from a JSON file". I'm thinking about whether this could be possible. Perhaps we could do a bit of parsing on the Beats side of the log filename to extract it's extension and set a field based on that. 🤔

**Comment by ycombinator** [comment]
> Another idea, possible stupid one.

> Try to parse the line as JSON and fallback to plain text.

Hmmm, I don't think it's stupid — this might actually work. Let me try this out!

**Comment by mtojek** [comment]
I admit I didn't dive deeper into libbeat, but maybe it's possible to sneak a format detection around the log file detection (similar to `http. DetectContentType`).

**Comment by ycombinator** [comment]
> Try to parse the line as JSON and fallback to plain text.

I tried this and it's effectively no different than detecting if the first character is `{` or not. That's because the real issue isn't so much with detecting JSON vs. not _in the Elasticsearch ingest pipeline_. The issue is detecting whether we're part of a plaintext message or not _on the Beats side_, before we send the event to Elasticsearch. 

In particular, we need to know if we're looking at JSON but it's part of a multiline plaintext event, in which case we should make sure that event is processed by the plaintext pipeline. For this, I think the solution has to come into play somewhere in the libbeat log reader: https://github.com/elastic/beats/blob/95626b8f1690344312c0831ab2bdcbccffe4d089/filebeat/input/log/harvester.go#L630.

I think we should explore this separately in a follow up issue/PR as it has the potential to impact not just the `logstash` Filebeat module. I'm keen to get this PR here merged as it will unblock moving this module to EPM (per my understanding of https://github.com/elastic/package-registry/pull/270#discussion_r393719035). WDYT?

---

BTW, I did find a way to make the "are we reading a JSON object as part of a multiline plaintext event" scenario _slightly_ more robust by changing the multiline `pattern` from:

```
  pattern: ^(\[[0-9]{4}-[0-9]{2}-[0-9]{2}|{)
```

to:

```
  pattern: ^(\[[0-9]{4}-[0-9]{2}-[0-9]{2}|({.+}))
```

It's not perfect (that would require a PDA whereas regexps are modeled as FSMs) but it's slightly more robust.

**Comment by mtojek** [comment]
>  I'm keen to get this PR here merged as it will unblock moving this module to EPM (per my understanding of elastic/package-registry#270 (comment)). WDYT?

Thank you for exploring other possibilities and describing the flow (TIL :) ) I'm not blocking this change at all, it's already an improvement. Ship it, please!

>   pattern: ^(\[[0-9]{4}-[0-9]{2}-[0-9]{2}|({.+}))

For completeness of the regex you can add the closing `]`. Also, maybe it's worth noting (even in docs) that the pipeline is not perfect, as it will break down for similar edge-cases:

```
{a} hello world {b}
```

**Comment by ycombinator** [comment]
> Also, maybe it's worth noting (even in docs) that the pipeline is not perfect, as it will break down for similar edge-cases.

Good idea, will add!

**Comment by ycombinator** [comment]
@mtojek Turns out it's not a matter of simply adding a closing `]` to the regex pattern. That is, we can't just change it from:

```
```

to:

```
```

That's because the timestamp pattern is incomplete in the regex. It only accounts for the date part, not the time part. So either we have to change the regex to:

```
^((\[[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2},[0-9]{3}\])|({.+}))
```

or to:

```
^((\[[0-9]{4}-[0-9]{2}-[0-9]{2}[^\]]+\])|({.+}))
```

I'm not sure either change, for the sake of completeness, is worth the extra processing. The purpose of this regex is simply to detect if a new multiline event should be started (and the previous one completed) or not. So I'm going to leave the regex as-is.

**Pull request merged**